HPCC-33333 In GH Unittest Action change the test case execution sequence and add core file check.

AttilaVamos opened the pull request:
Phase 1:
- Change the execution sequence and log the test case name

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented:
@Michael-Gardner please review this change.

Michael-Gardner reviewed (approved):
Attila this looks great.

AttilaVamos commented:
@ghalliday please merge

jakesmith commented on .github/workflows/test-unit-gh_runner.yml:
```diff
@@ -54,10 +57,42 @@ jobs:
+
+      - name: Check for Core files
+        run: |
+          CORE_FILES=( $(sudo find /opt/HPCCSystems/bin/ /var/lib/HPCCSystems/ -iname 'core*' -type f -print) )
```
are you sure these are the right directories to look in?

if unittest cores, it will dump in the current working director which will be the runner cwd afaics - i.e. there won't be any core files in /opt/HPCCSystems/bin/ or /var/lib/HPCCSystems/.

Also, is 'sudo' necessary?

If there can only be <=1 core and only from unittest in cwd then this could be simplified to:
```
    shopt -s nullglob
    CORE_FILES=(core*)
```

AttilaVamos replied:
Changed

jakesmith reviewed (commented):
@AttilaVamos - please see comments.

AttilaVamos commented on .github/workflows/test-unit-gh_runner.yml:
```diff
@@ -41,6 +43,7 @@ jobs:
           sudo apt-get update
           sudo apt-get install -y ${{ inputs.dependencies }}
           sudo apt-get install -y gdb
+          echo 'core_%e.%p' | sudo tee /proc/sys/kernel/core_pattern
```
OBT is using this kind of core creation setting in years now.

jakesmith replied:
ok, unrelated to this PR, but does the github smoketest action also need it to ensure cores/stacks created and upload to artifacts where too?

AttilaVamos commented on .github/workflows/test-unit-gh_runner.yml:
```diff
@@ -54,10 +57,42 @@ jobs:
+            echo "--------------------------------------"
+            sudo /opt/HPCCSystems/bin/unittests -e "$test"
+            echo " "
+          done< <(/opt/HPCCSystems/bin/unittests -l | sort )
```
Yes, the step fail at the first failed test.

AttilaVamos commented on .github/workflows/test-unit-gh_runner.yml:
```diff
@@ -54,10 +57,42 @@ jobs:
 
       - name: unittests-logs-artifact
-        if: ${{ failure() || cancelled() }}
+        if: ${{ failure() || cancelled() || env.UPLOAD_ARTIFACT == 'true'}}
```
True, removed.

jakesmith reviewed (approved):
@AttilaVamos - looks good. Please squash.

1 question re. core pattern/files/stacks in smoketest beyond scope of this PR.

jakesmith commented on .github/workflows/test-unit-gh_runner.yml:
```diff
@@ -52,9 +53,43 @@ jobs:
         timeout-minutes: 10 # generous, expected time is approx 1 min.
         shell: "bash"
         run: |
+          set +e     # Ignore intermediate error(s) in this step
```
Is this going to mean that it will no longer stop after 1st unittest failure, and could run through all (generating a core file for each potentially) ?

That may be okay, but what is the intent of the set +e here?

AttilaVamos replied:
If "Bash"specified, then "bash -e {0}" executed internally. You can claim it need to fail fast, then I will remove the 'set +e'.

AttilaVamos committed:
HPCC-33333 In GH Unittest Action change the testcase execution sequence and add core file check.
Phase 1:
- Change the executon sequence and log the test case name
- Add core check and core tracing generation code

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@ghalliday squashed. (You can see the unit test result: https://github.com/hpcc-systems/HPCC-Platform/actions/runs/13306096556/job/37158977242?pr=19469)

jakesmith reviewed (approved):
Re-approving now squashed. NB: all other reviewers had already approved - squashed commit introduces no new changes.
